examples/example-hono-api/src/auth0.ts

```diff
@@ -0,0 +1,175 @@
+      apiClient = new ApiClient({
+        domain: auth0_domain,
+        audience: auth0_audience,
+      });
+    }
+
+    const credentials = ctx.req.raw.headers.get('Authorization');
+    if (!credentials) {
+      const errDescription = 'No Authorization header included in request';
+      throw new HTTPException(401, {
+        message: errDescription,
+        res: unauthorizedResponse({
+          ctx,
+          error: 'invalid_request',
+          errDescription,
+        }),
+      });
+    }
+
+    const parts = credentials.split(/\s+/);
+    if (parts.length !== 2) {
+      const errDescription = 'Invalid Authorization header structure';
+      throw new HTTPException(401, {
+        message: errDescription,
+        res: unauthorizedResponse({
+          ctx,
+          error: 'invalid_request',
+          errDescription,
+        }),
+      });
+    }
+
+    if (parts[0] !== 'Bearer') {
+      const errDescription =
+        'Invalid authorization header (only Bearer tokens are supported)';
+      throw new HTTPException(401, {
+        message: errDescription,
+        res: unauthorizedResponse({
+          ctx,
+          error: 'invalid_request',
+          errDescription: errDescription,
+        }),
+      });
+    }
+
+    const accessToken = parts[1];
+    if (!accessToken || accessToken.length === 0) {
+      const errDescription = 'No token included in request';
+      throw new HTTPException(401, {
+        message: errDescription,
+        res: unauthorizedResponse({
+          ctx,
+          error: 'invalid_request',
+          errDescription,
+        }),
+      });
+    }
+
+    try {
+      const jwtPayload = await apiClient.verifyAccessToken({ accessToken });
+
+      ctx.set('jwtPayload', jwtPayload);
+
+      await next();
+    } catch (cause) {
+      console.trace(cause);
+
+      throw new HTTPException(401, {
+        message: 'Unauthorized',
+        res: unauthorizedResponse({
+          ctx,
+          error: 'invalid_token',
+          statusText: 'Unauthorized',
+          errDescription: 'Token verification failure',
+        }),
+        cause,
+      });
+    }
+
+  };
+};
+
+function unauthorizedResponse(opts: {
+  ctx: Context;
+  error: string;
+  errDescription: string;
+  statusText?: string;
+}) {
+  return new Response('Unauthorized', {
+    status: 401,
+    statusText: opts.statusText,
+    headers: {
+      'WWW-Authenticate': `Bearer realm="${opts.ctx.req.url}",error="${opts.error}",error_description="${opts.errDescription}"`,
+    },
+  });
+}
+
+export function requireScope(scope: string | string[]): MiddlewareHandler {
+
+  return async function requireScope(ctx, next) {
+    const payload = ctx.var.jwtPayload as JWTPayload;
+
+    if (!payload.scope) {
+      throw new HTTPException(403, {
+        message: 'Forbidden',
+        res: unauthorizedResponse({
+          ctx,
+          error: 'insufficient_scope',
+          errDescription: `Missing required scope: ${scope}`,
+        }),
```

Comment:
As far as I know, `insufficient_scope` shouldnt be a 401 , any reason we are doing this here?

Reply:
Additionally, I feel like this is a lot of code to validate the scopes (we even throw the same error twice), while we use this in all other samples and SDKs:

```
function validateScopes(token: Token, requiredScopes: string[]): boolean {
  let tokenScopes: string[] = [];

  if (token.scope) {
    tokenScopes =
      typeof token.scope === 'string' ? token.scope.split(' ') : token.scope;
  }

  return requiredScopes.every((required) => tokenScopes.includes(required));
}
```

I feel like using this could simplify things.